src/main/kotlin/no/elhub/devxp/autochangelog/AutoChangelog.kt

```diff
@@ -0,0 +1,119 @@
+
+    @CommandLine.Option(
+        names = ["--from-tag", "--from"],
+        required = false,
+        description = ["The tag to start the changelog from (exclusive)"]
+    )
+    var fromTagName: String? = null
+
+    @CommandLine.Option(
+        names = ["--to-tag", "--to"],
+        required = false,
+        description = ["The tag to end the changelog at (inclusive)"]
+    )
+    var toTagName: String? = null
+
+    @CommandLine.Option(
+        names = ["-j", "-json"],
+        required = false,
+        description = ["Whether to output changelog in JSON format instead of Markdown"]
+    )
+    var json: Boolean = false
+
+    @CommandLine.Option(
+        names = ["--group-by-commit"],
+        required = false,
+        description = ["Whether to group the changelog entries by commit instead of JIRA issue"]
+    )
+    var commitGrouping: Boolean = false
+
+
+    override fun run() {
+        val gitRepository = initRepository(workingDir)
+
+        val tags = getTagsFromRepo(gitRepository)
+
+        val maybeFromTag = tags.firstOrNull { it.name == fromTagName }
+        val maybeToTag = tags.firstOrNull { it.name == toTagName }
+
+        val relevantCommits = getRelevantCommits(gitRepository, maybeFromTag, maybeToTag, tags)
+
+        // Create a mapping of JIRA issue IDs to commits
+        val jiraIssueIds = extractJiraIssuesIdsFromCommits(relevantCommits)
+
+        // Populate the map with actual JIRA issue details
+        val jiraMap: Map<JiraIssue, List<GitCommit>>
+        runBlocking {
+            jiraMap = client.populateJiraMap(jiraIssueIds)
+        }
+
+        val changelogName = "CHANGELOG"
+        val changeLogFileSuffix = if (maybeFromTag != null || maybeToTag != null) {
+            " [${maybeFromTag?.name ?: ""}-${maybeToTag?.name ?: ""}]"
+        } else {
+            ""
+        }
+
+        if (commitGrouping) {
+            // Reverse the mapping to be GitCommit -> List<JiraIssue>
+            val commitMap = jiraMap
+                .flatMap { (key, values) -> values.map { it to key } }
+                .groupBy({ it.first }, { it.second })
+                .toSortedMap((compareByDescending { it.commitTime }))
+
+            if (json) {
+                val jsonContent = formatCommitJson(commitMap)
+                writeJsonToFile(jsonContent, "$changelogName$changeLogFileSuffix.json")
+            } else {
+                val markdownContent = formatCommitMarkdown(commitMap)
+                writeMarkdownToFile(markdownContent, "$changelogName$changeLogFileSuffix.md")
+            }
+
+        } else {
+            if (json) {
+                val jsonContent = formatJson(jiraMap)
+                writeJsonToFile(jsonContent, "$changelogName$changeLogFileSuffix.json")
+            } else {
+                val markdownContent = formatMarkdown(jiraMap)
+                writeMarkdownToFile(markdownContent, "$changelogName$changeLogFileSuffix.md")
+            }
+        }
```

Comment:
:art: I think this could be rewritten/restructured to only need to invoke a single `writeToFile(content, FileFormat.Json)` (or `FileFormat.Markdown`) method once